### ISSUE TITLE
Pipenv 2020 error's out not finding the requirement.txt

### ISSUE THREAD

**Issue by saisankargochhayat** [pypa/pipenv#4239]
### Issue description
On a empty folder installing - 
`pipenv install thoth-sourcemanagement`
Error's out stating missing requirements.txt
However works as expected with pipenv 2018 version. 

### Steps to replicate
Environment - 
Fedora 32
You can also replicate the error on a katacoda environment - https://www.katacoda.com/courses/ubuntu/playground1804

Try installing the following Pipfile with pipenv 2020
Provide the steps to replicate (which usually at least includes the commands and the Pipfile).
```
[[source]]
name = "pypi"
url = "https://pypi.org/simple"
verify_ssl = true

[dev-packages]

[packages]
thoth-sourcemanagement = "*"

[requires]
python_version = "3.6"
```
![image](https://user-images.githubusercontent.com/9396452/81089875-ea14ef80-8eca-11ea-8dbf-da01919c9570.png)

**Comment by techalchemy** [comment]
The new version of pipenv upgrades the internal version of pip, which in turn relies on `pep517` based building and, if this worked in the past, apparently it handles errors a bit less gracefully.

The new setup doesn't make any promises about putting the package directory on `sys.path` at build time, so trying to open `requirements.txt` (as opposed to `os.path.join(os.path.dirname(os.path.abspath(__file__)), "requirements.txt")` for example) is not guaranteed to work. This is a departure in terms of build process but most projects felt the pain of this some time ago.

That said, this isn't the best practice for telling `setuptools` about your requirements, and it's not really supported in any of the tooling anymore so I'm not sure we will want to make substantial changes to accommodate this workflow

**Comment by taybin** [comment]
I've run into this as well.  Is the fix basically to migrate to using pyproject.toml instead of or in addition to setup.py for declaring dependencies?

**Comment by saisankargochhayat** [comment]
> I've run into this as well. Is the fix basically to migrate to using pyproject.toml instead of or in addition to setup.py for declaring dependencies?

@taybin So pyproject.toml is what will be used to build the eggs for pypi?

**Comment by taybin** [comment]
> @taybin So pyproject.toml is what will be used to build the eggs for pypi?

I don't know.  It's not like there's much documentation around on how to use this stuff.  Most documentation I see like (https://docs.python-guide.org/writing/structure/) doesn't mention pyproject.toml at all.  So I guess I need to use poetry for libraries, and pipenv for applications until I can switch them over to use poetry as well for consistency.  I'm just guessing though.

**Comment by techalchemy** [comment]
The fix is to not require **implicit** path lookups in your `setup.py`. 

The *real* fix is to declare dependencies directly in your `setup.py` or `setup.cfg`, the issue here has absolutely nothing to do with `pyproject.toml` whatsoever. The problem is that the python packaging ecosystem has relied on *quirks* of the packaging tooling that are not promises, like attempting to read a naked, relative path (in this case `requirements.txt`) because of the assumption that, at build time, the actual build utility will execute in the package's directory.

The problem is that this is not the case anymore -- builds don't have to happen in your package directory, they can take place from any directory. So, if I am in `/tmp/some-temp-build-dir` and I run a build against `/home/yourhomedir/yoursourcecode`, I'm still in `/tmp/some-temp-build-dir`.

The original post references `thoth-sourcemanagement` so I'll use their `setup.py` as the example here: https://github.com/thoth-station/source-management/blob/e7d044ffa5725bfde8774fb2556f41f2cca36fd4/setup.py#L22-L26

This is a direct excerpt from the package setup.py, which refers to `requirements.txt`. This is a path that needs to exist *relative to the current working directory of the build*. That is just not going to be true, because the build is happening in a temporary directory.

You can fix this with an absolute path, but you will still break dependency resolvers. You *SHOULD* fix this by specifying requirements in `setup.py` or `setup.cfg` instead of dynamically calculating them at runtime. If you're already using `setuptools` there is absolutely no reason to change all of your infrastructure just to specify dependencies statically, since `setuptools` obviously has that capability already.

This is widely misunderstood which I get, but while it's a little bit nuanced, the fix isn't that complicated. More importantly, these issues wouldn't happen in the first place with static declarations which would be required in `pyproject.toml` anyway.

**Comment by taybin** [comment]
> The _real_ fix is to declare dependencies directly in your `setup.py` or `setup.cfg`, the issue here has absolutely nothing to do with `pyproject.toml` whatsoever.

Thank you for the explanation.  This makes sense.  I'm confused why it stopped working and why using requirement.txt files has been so common if it's inherently broken.  A mismatch of the workflows for applications vs libraries?

**Comment by techalchemy** [comment]
> > The _real_ fix is to declare dependencies directly in your `setup.py` or `setup.cfg`, the issue here has absolutely nothing to do with `pyproject.toml` whatsoever.
> 
> Thank you for the explanation. This makes sense. I'm confused why it stopped working and why using requirement.txt files has been so common if it's inherently broken. A mismatch of the workflows for applications vs libraries?

The reason people make the association between `pyproject.toml` and the issue described here is because both changes occurred with the same release of `pip` and `setuptools` (and coincidentally marked the start of a series of delays in being able to put together a release of `pipenv`) -- `pep517` was finally implemented as a default behavior in `pip`'s build process in december 2018 I think, and by default if projects had a `pyproject.toml` present, `pip` would assume they meant to use the new build process.

Firstly, this didn't always work because the mere presence of the file doesn't mean that the requisite keys have been defined (`[build-sustem.requires]` and `[build-system.build-backend]`), and secondly because of the issue described above -- the previous build process never left the package directory to perform the build, so many times people would assume that they could use a relative path in `setup.py` or a relative import (e.g. importing their own library) and this would work just fine.

With the isolated `pep517` build process, the package gets cloned to a temporary directory and the build launched inside a subprocess, but the temporary directory is not added to `sys.path` and the subprocess is not set to `chdir` to that directory before executing the build.

Basically, this broke a lot of assumptions, and the outcome of a number of discussions about the topic was that `setuptools` and `pip` wouldn't be accommodating this use case anymore as it shouldn't have worked to begin with.

I'm not a fan of breaking backward compatibility, but I have to balance that with supporting the modern build processes as well so we are basically stuck with this as the current situation. Statically defined requirements are easier to resolve anyway so it will likely be better in the long run even if it is more painful short term.

**Comment by taybin** [comment]
FWIW, I can't duplicate this anymore, even without updating my libraries to not use a requirements.txt file.  Still planning on doing that, but not having my builds break is also a relief. 👍

**Comment by techalchemy** [comment]
I did a bit of extra work to try parsing `setup.py` without breaking, and it seems to have paid off in your case. For this case and others like it, pipenv falls back to `requirementslib` which ultimately tries to build a wheel using the `pep517` wheel builder and, if unspecified, the `setuptools` legacy build backend. If that is successful we will consume metadata from the built wheel -- this is obviously much less efficient than just parsing textual metadata that is declared in a manifest, but it's kind of neat that it works again ^^

I'll close this issue for now as I was able to confirm this myself directly with `requirementslib`:

```python
>>> from requirementslib.models.requirements import Requirement                                                                                            
>>> r = Requirement.from_line("-e .")                                                                                                                      
>>> r.run_requires()                                                                                                                                       
{'name': 'thoth-sourcemanagement', 'version': '0.2.9', 'base_dir': '/tmp/reqlib-srczugi9sxi/source-management', 'ireq': <InstallRequirement object: thoth-sourcemanagement==0.2.9@ file:///tmp/source-management from file:///tmp/source-management editable=True>, 'build_backend': 'setuptools.build_meta:__legacy__', 'build_requires': ('wheel', 'setuptools>=40.8'), 'requires': {'ogr': Requirement.parse('ogr'), 'requests': Requirement.parse('requests'), 'delegator.py': Requirement.parse('delegator.py')}, 'python_requires': '>=3.6', 'extra_kwargs': {'build_dir': '/tmp/reqlib-buildyrhlbxd0', 'src_dir': '/tmp/reqlib-srczugi9sxi', 'download_dir': '/home/techalchemy/.cache/pipenv/pkgs', 'wheel_download_dir': '/home/techalchemy/.cache/pipenv/wheels'}, 'setup_cfg': PosixPath('/tmp/reqlib-srczugi9sxi/source-management/setup.cfg'), 'setup_py': PosixPath('/tmp/reqlib-srczugi9sxi/source-management/setup.py'), 'pyproject': PosixPath('/tmp/reqlib-srczugi9sxi/source-management/pyproject.toml')}
```
